Aplib.Core/ExposedQueue.cs

```diff
@@ -0,0 +1,193 @@
+{
+    /// <summary>
+    /// A queue with all elements exposed.
+    /// Functionally works like a queue with indexing.
+    /// </summary>
+    public class ExposedQueue<T> : ICollection<T>
+    {
+        /// <summary>
+        /// The length of the array.
+        /// </summary>
+        public int MaxCount { get; private set; }
+
+        /// <summary>
+        /// Actual number of elements in the array.
+        /// Limited to <see cref="MaxCount"/>.
+        /// </summary>
+        public int Count
+        {
+            get => _count;
+            private set => _count = value > MaxCount ? MaxCount : value;
+        }
+
+        /// <inheritdoc/>
+        public bool IsReadOnly => false;
+
+        private int _count;
+
+        private readonly T[] _array;
+        private int _head;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ExposedQueue{T}"/> class.
+        /// </summary>
+        /// <param name="size">The size of the array.</param>
+        public ExposedQueue(int size)
+        {
+            MaxCount = size;
+            _array = new T[MaxCount];
+            _head = MaxCount - 1;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ExposedQueue{T}"/> class.
+        /// By default, assumes the array is filled.
+        /// </summary>
+        /// <param name="array">An array to use as the circular array.</param>
+        /// <param name="count">The number of actual elements in the array.</param>
+        public ExposedQueue(T[] array, int count = -1)
+        {
+            MaxCount = array.Length;
+            _array = array;
+            _head = MaxCount - 1;
+            Count = count == -1 ? MaxCount : count;
+        }
+
+        /// <summary>
+        /// Gets the element at the specified index.
+        /// </summary>
+        /// <param name="index">The index of the element to get.</param>
+        /// <returns>The element at the specified index.</returns>
+        public T this[int index]
+        {
+            get => _array[(index + _head + 1) % MaxCount];
+            private set => _array[(index + _head + 1) % MaxCount] = value;
+        }
+
+        /// <summary>
+        /// Decrements the head of the array.
+        /// </summary>
+        private void DecrementHead() => _head = (_head - 1 + MaxCount) % MaxCount;
+
+        /// <summary>
+        /// Puts an element at the start of the array.
+        /// </summary>
+        /// <param name="value">The element to add to the array</param>
+        public void Put(T value)
+        {
+            _array[_head] = value;
+            DecrementHead();
+            Count++;
+        }
+
+        /// <inheritdoc/>
+        public void Add(T item) => Put(item);
+
+        /// <summary>
+        /// Gets the element at the head of the array.
+        /// </summary>
+        /// <returns>The element at the head of the array</returns>
+        public T GetHead() => _array[_head];
```

Comment:
Could this be renamed to GetLast? 'Head' could be interpreted in various ways, while GetLast is less ambiguous semantically. This remark is just for the publically visible GetHead.
```suggestion
        public T GetLast() => _array[_head];
```